[WebGPU] Implement RenderPipeline::getBindGroupLayout

mwyrzykowski opened the pull request:
#### d0ba6e761c62aeca09022fdbccc2b73f17fa0d08
<pre>
[WebGPU] Implement RenderPipeline::getBindGroupLayout
<a href="https://bugs.webkit.org/show_bug.cgi?id=249539">https://bugs.webkit.org/show_bug.cgi?id=249539</a>
&lt;radar://103482448&gt;

Reviewed by Myles C. Maxfield.

Use reflection to build the bind group layout for a given
bind group from the render pipeline.

* Source/WebGPU/WebGPU/BindGroupLayout.h:
(WebGPU::BindGroupLayout::create):
(WebGPU::BindGroupLayout::createInvalid):
(WebGPU::BindGroupLayout::device const): Deleted.
Device isn&apos;t used and not directly accessible from RenderPipeline
member functions so remove it as BindGroupLayout is created
in the RenderPipeline now.

* Source/WebGPU/WebGPU/BindGroupLayout.mm:
(WebGPU::Device::createBindGroupLayout):
(WebGPU::BindGroupLayout::BindGroupLayout):
Be more specific about the value type in the key-value pair.

* Source/WebGPU/WebGPU/RenderPipeline.h:
(WebGPU::RenderPipeline::create):
Pass the reflection info to be used in getBindGroupLayout.

* Source/WebGPU/WebGPU/RenderPipeline.mm:
(WebGPU::Device::createRenderPipeline):
(WebGPU::RenderPipeline::RenderPipeline):
(WebGPU::RenderPipeline::getBindGroupLayout):
Build a bind group layout with required binding index and stage
visibility for the render pipeline.

* Websites/webkit.org/demos/webgpu/scripts/hello-cube-auto-layout.js:
* Websites/webkit.org/demos/webgpu/scripts/instanced-textured-cube.js:
Update two samples illustrating the use of getBindGroupLayout

Canonical link: <a href="https://commits.webkit.org/258524@main">https://commits.webkit.org/258524@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/83f6c7420f297ab1de9bc819bb5c4cd5e183478b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/102195 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/11339 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/35264 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/111518 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/171680 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/12319 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/2253 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/94562 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/109247 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/107975 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/9437 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/92711 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/37229 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/91305 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/24193 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/78964 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/92541 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/4883 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/25616 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/88760 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/2527 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/5003 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/2054 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/29498 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/11049 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/45113 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/91681 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5842 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/6750 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/20492 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

litherum commented on Source/WebGPU/WebGPU/RenderPipeline.mm:
```diff
@@ -445,9 +446,10 @@ static MTLVertexStepFunction stepFunction(WGPUVertexStepMode stepMode)
 }
 
-RenderPipeline::RenderPipeline(id<MTLRenderPipelineState> renderPipelineState, MTLPrimitiveType primitiveType, std::optional<MTLIndexType> indexType, MTLWinding frontFace, MTLCullMode cullMode, id<MTLDepthStencilState> depthStencilState, uint32_t vertexShaderInputBufferCount, Device& device)
+RenderPipeline::RenderPipeline(id<MTLRenderPipelineState> renderPipelineState, MTLRenderPipelineReflection *reflection, MTLPrimitiveType primitiveType, std::optional<MTLIndexType> indexType, MTLWinding frontFace, MTLCullMode cullMode, id<MTLDepthStencilState> depthStencilState, uint32_t vertexShaderInputBufferCount, Device& device)
```
We shouldn't need Metal reflection to build a bind group layout. The rules about how to do it should be rigorously defined in WebGPU + WGSL, not Metal.

mwyrzykowski replied:
Yes we can get this information from the WGSL compiler when it is further along. For auto layout I would expect the generated layout from the WGSL compiler to match the reflection information from Metal, otherwise when building the argument buffer the encoding will not match. 

But really this path should only be triggered when the pipeline is created with auto layout. Otherwise we should return the layout passed to create pipeline. I will address that.

litherum commented on Source/WebGPU/WebGPU/BindGroup.mm:
```diff
@@ -50,6 +50,35 @@ static bool textureViewIsPresent(const WGPUBindGroupEntry& entry)
     return entry.textureView;
 }
 
+#if HAVE(TIER2_ARGUMENT_BUFFERS)
```
How can this be a preprocessor macro? It's hardware-dependent, not known at build time.

Also, why do we need any feature from tier 2 argument buffers at all? They're supposed to be able to work just fine with tier 1 only.

mwyrzykowski replied:
Naming of `TIER2_ARGUMENT_BUFFERS` could be improved, it is more something along the lines of `HAVE(SUPPORTS_GPU_ADDRESS)`. I check the tier level inside the `#if` prior to building a tier 2 buffer.

We certainly don't need tier 2 argument buffers. I used it to simplify the implementation until we want to run this on pre-A13 iOS device (iPhone 8, iPhone X) or an Intel Mac with non-discrete GPU and I work out logic errors in building the argument buffers from the GPUBindGroupDescriptors

mwyrzykowski replied:
Most impacted in the pre-A13 category are iPads I think. I certainly think we need to support tier1 argument buffers, I plan to add support once I address remaining issues in the bind group implementation.

litherum reviewed (commented):
Will dig in deeper later today

djg commented on Source/WebGPU/WebGPU/RenderPipeline.mm:
```diff
@@ -445,14 +450,33 @@ static MTLVertexStepFunction stepFunction(WGPUVertexStepMode stepMode)
+#endif
+}
+
+RenderPipeline::RenderPipeline(id<MTLRenderPipelineState> renderPipelineState, MTLPrimitiveType primitiveType, std::optional<MTLIndexType> indexType, MTLWinding frontFace, MTLCullMode cullMode, id<MTLDepthStencilState> depthStencilState, const PipelineLayout &pipelineLayout, uint32_t vertexShaderInputBufferCount, Device& device)
```
This looks almost identical to the previous decl that I went cross-eyed trying to see the different between them.

djg reviewed (approved):
LGTM but defer to @litherum

litherum commented on Source/WTF/wtf/PlatformHave.h:
```diff
@@ -1138,6 +1138,7 @@
 
 #if HAVE(WEBGPU_IMPLEMENTATION) && ((PLATFORM(MAC) && __MAC_OS_X_VERSION_MIN_REQUIRED >= 130000) || (PLATFORM(IOS) && __IPHONE_OS_VERSION_MIN_REQUIRED >= 160000))
 #define HAVE_TIER2_ARGUMENT_BUFFERS 1
+#define HAVE_METAL_BUFFER_BINDING_REFLECTION 1
```
I guess this is all throwaway code anyway, but I think reflection is available way further back than this macro indicates.

mwyrzykowski replied:
Yes, but the reflection APIs used are new to iOS 16 / macOS 13, the older ones are deprecated

litherum commented on Source/WebGPU/WebGPU/BindGroup.mm:
```diff
@@ -58,14 +87,18 @@ static bool textureViewIsPresent(const WGPUBindGroupEntry& entry)
+        constexpr auto shaderStageLength = WTF_ARRAY_LENGTH(shaderStage);
+        id<MTLBuffer> argumentBuffer[shaderStageLength];
+        for (size_t i = 0; i < shaderStageLength; ++i)
+            argumentBuffer[i] = bufferSizes[i] ? safeCreateBuffer(bufferSizes[i], MTLStorageModeShared) : nil;
```
I still think we shouldn’t be manually calculating byte sizes ourself but should be using MTLArgumentEncoder.encodedLength

litherum commented on Source/WebGPU/WebGPU/BindGroup.mm:
```diff
@@ -58,14 +87,18 @@ static bool textureViewIsPresent(const WGPUBindGroupEntry& entry)
+            argumentBuffer[i] = bufferSizes[i] ? safeCreateBuffer(bufferSizes[i], MTLStorageModeShared) : nil;
+
+        char* argumentBufferContents[] = { (char*)argumentBuffer[0].contents, (char*)argumentBuffer[1].contents, (char*)argumentBuffer[2].contents };
+        size_t bindingOffsetForStage[] = { 0, 0, 0 };
```
Where does this get populated?

mwyrzykowski replied:
`auto& offset = bindingOffsetForStage[currentStage];` on line 119

and then `offset += ` on lines 129, 134, 140

litherum commented on Source/WebGPU/WebGPU/BindGroup.mm:
```diff
@@ -77,34 +110,39 @@ static bool textureViewIsPresent(const WGPUBindGroupEntry& entry)
+                    if (entry.offset > buffer.length)
+                        return BindGroup::createInvalid(*this);
+
+                    ASSERT(sizeof(float*) == sizeof(buffer.gpuAddress));
```
Ditto about using MTLArgumentEncoder instead of gpuAddress. Might as well do it once in a cross-version way instead of having to do it twice

mwyrzykowski replied:
I agree, I filed https://bugs.webkit.org/show_bug.cgi?id=250123 which I will get to soon, I did it this way since it seemed simpler at the time

litherum commented on Source/WebGPU/WebGPU/BindGroupLayout.h:
```diff
@@ -41,61 +41,43 @@ class Device;
-    BindGroupLayout(id<MTLBuffer> vertexArgumentBuffer, id<MTLBuffer> fragmentArgumentBuffer, id<MTLBuffer> computeArgumentBuffer, HashMap<uint32_t, uint32_t>&&, Device&);
-    BindGroupLayout(Device&);
-
-    union {
```
I don’t know what these unions were there for, but I’m glad to see them gone 😅

litherum commented on Source/WebGPU/WebGPU/RenderPipeline.h:
```diff
@@ -36,14 +37,20 @@ namespace WebGPU {
     WTF_MAKE_FAST_ALLOCATED;
 public:
-    static Ref<RenderPipeline> create(id<MTLRenderPipelineState> renderPipelineState, MTLPrimitiveType primitiveType, std::optional<MTLIndexType> indexType, MTLWinding frontFace, MTLCullMode cullMode, id<MTLDepthStencilState> depthStencilState, uint32_t vertexShaderInputBufferCount, Device& device)
+    static Ref<RenderPipeline> create(id<MTLRenderPipelineState> renderPipelineState, MTLPrimitiveType primitiveType, std::optional<MTLIndexType> indexType, MTLWinding frontFace, MTLCullMode cullMode, id<MTLDepthStencilState> depthStencilState, MTLRenderPipelineReflection *reflection, uint32_t vertexShaderInputBufferCount, Device& device)
```
I’m reading your previous comment to mean that this is throwaway code that will be replaced soon

litherum replied:
Also, a bind group layout contains more information than metal reflection, so the sooner we can do this using WebGPU types, the better

mwyrzykowski replied:
Yes, it would be great to get this information directly from the WGSL compiler

litherum commented on Source/WebGPU/WebGPU/RenderPipeline.mm:
```diff
@@ -445,14 +450,33 @@ static MTLVertexStepFunction stepFunction(WGPUVertexStepMode stepMode)
+    : m_renderPipelineState(renderPipelineState)
+    , m_device(device)
+    , m_primitiveType(primitiveType)
+    , m_indexType(indexType)
```
Wow this patch is doing a lot

mwyrzykowski replied:
It does build on https://github.com/WebKit/WebKit/pull/8042

only top commit needs review after the first one is merged https://github.com/WebKit/WebKit/commit/17cd5f800c0db82254db1f0bed578f0e9c094616

litherum commented on Websites/webkit.org/demos/webgpu/scripts/hello-cube-auto-layout.js:
```diff
@@ -76,20 +76,6 @@ async function helloCube() {
     });
 
     /*** Shader Setup ***/
-    
```
I think modifying these files will actually push updates to the real live site. Is that really what we want?

mwyrzykowski replied:
It should be fine, since the feature is still disabled at runtime so we will see the same behavior on Safari and STP as before (WebGPU is not supported). Unless we don’t want the files accessible, but I think you can only reach many of the html files if you were to know the full path as the index.html file is unchanged.

litherum reviewed (approved):
I have some concerns, but we’re rapidly developing, so I guess it’s ok and we can go from here

mwyrzykowski committed:
[WebGPU] Implement RenderPipeline::getBindGroupLayout
https://bugs.webkit.org/show_bug.cgi?id=249539
<radar://103482448>

Reviewed by Myles C. Maxfield.

Use reflection to build the bind group layout for a given
bind group from the render pipeline.

* Source/WebGPU/WebGPU/BindGroupLayout.h:
(WebGPU::BindGroupLayout::create):
(WebGPU::BindGroupLayout::createInvalid):
(WebGPU::BindGroupLayout::device const): Deleted.
Device isn't used and not directly accessible from RenderPipeline
member functions so remove it as BindGroupLayout is created
in the RenderPipeline now.

* Source/WebGPU/WebGPU/BindGroupLayout.mm:
(WebGPU::Device::createBindGroupLayout):
(WebGPU::BindGroupLayout::BindGroupLayout):
Be more specific about the value type in the key-value pair.

* Source/WebGPU/WebGPU/RenderPipeline.h:
(WebGPU::RenderPipeline::create):
Pass the reflection info to be used in getBindGroupLayout.

* Source/WebGPU/WebGPU/RenderPipeline.mm:
(WebGPU::Device::createRenderPipeline):
(WebGPU::RenderPipeline::RenderPipeline):
(WebGPU::RenderPipeline::getBindGroupLayout):
Build a bind group layout with required binding index and stage
visibility for the render pipeline.

* Websites/webkit.org/demos/webgpu/scripts/hello-cube-auto-layout.js:
* Websites/webkit.org/demos/webgpu/scripts/instanced-textured-cube.js:
Update two samples illustrating the use of getBindGroupLayout

Canonical link: https://commits.webkit.org/258524@main
